contracts/Governance.sol

```diff
@@ -0,0 +1,256 @@
+        proposal.quorumToken = MiniMeToken(cloneToken);
+        proposal.quorumToken.changeController(tokenController);
+
+        proposal.state = ProposalState.PutToSupport;
+
+        emit ProposalPutToSupport(_id);
+    }
+
+
+    /** @dev Calculate and return required quorum for a given proposal.
+     *  @param _id ID of a proposal
+     */
+    function getRequiredQuorum(bytes32 _id) public view onlyWhenProposalPutToSupport(_id) returns (uint effectiveQuorum){
+        uint numberOfDividePeriodsPassed = (block.timestamp - lastTimeQuorumReached) / quorumDivideTime;
+        effectiveQuorum = currentProposalQuorum * proposals[_id].quorumToken.totalSupply() / (2 ** numberOfDividePeriodsPassed) / 100;
+    }
+
+
+    /** @dev Put given proposal to vote.
+     *  @param _id ID of a proposal
+     */
+    function putProposalToVote(bytes32 _id) public onlyWhenProposalPutToSupport(_id) {
+        require(proposals[_id].quorumToken.balanceOf(supportDeposit) >= getRequiredQuorum(_id), "Only when propsal has quorum.");
+
+        Proposal storage proposal = proposals[_id];
+
+        proposal.whenPutToVote = block.timestamp;
+
+        address cloneToken = pinakion.createCloneToken({_cloneTokenName: voteTokenName, _cloneDecimalUnits: DECIMALS, _cloneTokenSymbol: voteTokenSymbol, _snapshotBlock: block.number, _transfersEnabled: true});
+        proposal.voteToken = MiniMeToken(cloneToken);
+        proposal.voteToken.changeController(tokenController);
+
+        proposal.state = ProposalState.PutToVote;
+
+        emit ProposalPutToVote(_id);
+
+        lastTimeQuorumReached = block.timestamp; // Necessary when calculating required quorum as it is halved periodically.
+        currentProposalQuorum = proposalQuorum; // Update required quorum percent, which will be constant during new quorum phase.
+        currentVotingTime = votingTime; // Update allowed voting time, which will be constant during new quorum phase.
+    }
+
+
+    /** @dev Ends a voting, moves proposal to decided state, sets the decision
+     *  @param _id ID of a proposal
+     */
+    function finalizeVoting(bytes32 _id) public  {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        require(now - proposals[_id].whenPutToVote >= currentVotingTime, "Only when voting period passed.");
+
+        proposals[_id].state = ProposalState.Decided;
+        proposals[_id].approved = proposals[_id].voteToken.balanceOf(approvalDeposit) > proposals[_id].voteToken.balanceOf(rejectionDeposit);
+
+        emit ProposalDecided(_id, proposals[_id].approved);
+    }
+
+
+    /** @dev General purpose call function for executing a proposal UNTRUSTED
+     *  @param _id ID of a proposal
+     */
+    function executeProposal(bytes32 _id) public {
+        require(proposals[_id].state == ProposalState.Decided, "Only when proposal in Decided state.");
+        require(proposals[_id].approved, "Only when proposal is approved.");
```

Comment:
I would change the wording on all of these exception messages from "Only when..." to "... is not ..." or "... must be ....". Like `Proposal is not approved` and `Proposal must be in Decided state`